tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ToEtcExpr.scala

```diff
@@ -627,23 +627,59 @@ class ToEtcExpr(annotationStore: AnnotationStore, varPool: TypeVarPool) extends
           accessorsWithNewValues.flatMap(p => List(this(p._1), this(p._2)))
         mkApp(ref, disjunctiveType, this(fun) +: xargs: _*)
 
-      case OperEx(TlaFunOper.recFunDef, body, NameEx(name), bindingSet) =>
-        // the expected type is: (((a -> b) => (a => b)) => (a -> b)) (λ $recFun ∈ Set(c -> d). λ x ∈ Int. x)
+      case OperEx(TlaFunOper.recFunDef, body, args @ _*) =>
+        // fun[x \in S, y \in T |-> e] == ...
+        // or, fun[<<x, y>> \in S, z \in T |-> e] == ...
+        //
+        // The expected type is for a one-argument function is:
+        // (((b -> a) => (b => a)) => (b -> a)) (λ $recFun ∈ Set(c -> d). λ x ∈ Int. x)
+        //
+        // (The case of multiple arguments is quite complex)
+        val bindings =
+          translateBindings(
+              args
+                .grouped(2)
+                .map {
+                  case Seq(varEx, setEx) => (varEx, setEx)
+                  case orphan            => throw new TypingException(s"Invalid bound variables and sets ${orphan} in: ${ex}")
+                }
+                .toSeq: _*
+          )
+
         val a = varPool.fresh
-        val b = varPool.fresh
-        val funType = FunT1(a, b)
-        val aToB = OperT1(Seq(a), b)
-        val principal = OperT1(Seq(OperT1(Seq(funType), aToB)), funType)
+        // start with "b", as "a" goes to the result
+        val typeVars = varPool.fresh(bindings.length)
+
+        val funFrom = typeVars match {
+          // With one argument, the generated function has the type b -> a, that is, no tuple is involved.
+          case Seq(one) => one
+          // With multiple arguments, the generated function has the type <<b, c>> -> a, that is, it accepts a tuple
+          case many => TupT1(many: _*)
+        }
+
+        // e.g., b -> a, or <<b, c>> -> a
+        val funType = FunT1(funFrom, a)
+        // e.g., b => a, or (b, c) => a
+        val operType = OperT1(typeVars, a)
+        val principal = OperT1(Seq(OperT1(Seq(funType), operType)), funType)
+        // λ x ∈ S, y ∈ T. [[body]]
         val innerLambda =
-          mkAbs(ExactRef(body.ID), this(body), (name, this(bindingSet)))
-        val c = varPool.fresh
-        val d = varPool.fresh
+          mkAbs(ExactRef(body.ID), this(body), bindings: _*)
+        // create another vector of type variables for the lambda over a function
+        val recFunResTypeVar = varPool.fresh
+        // similar to funFrom
+        val recFunFrom = varPool.fresh(bindings.length) match {
+          // With one argument, the generated function has the type d -> c, that is, no tuple is involved.
+          case Seq(one) => one
+          // With multiple arguments, the generated function has the type <<d, e>> -> c, that is, it accepts a tuple
+          case many => TupT1(many: _*)
+        }
```

Comment:
Seems like this is general enough it should be its own auxiliary function. Then you wouldn't have to repeat the comments :)